x/staking/handler.go

```diff
@@ -229,7 +229,17 @@ func handleMsgDelegate(ctx sdk.Context, msg types.MsgDelegate, k keeper.Keeper)
 }
 
 func handleMsgUndelegate(ctx sdk.Context, msg types.MsgUndelegate, k keeper.Keeper) sdk.Result {
-	completionTime, err := k.Undelegate(ctx, msg.DelegatorAddress, msg.ValidatorAddress, msg.SharesAmount)
+	validator, found := k.GetValidator(ctx, msg.ValidatorAddress)
+	if !found {
+		return ErrNoValidatorFound(k.Codespace()).Result()
+	}
+
+	shares, err := validator.SharesFromTokens(msg.Amount.Amount)
+	if err != nil {
+		return err.Result()
+	}
```

Comment:
We should also be capping the shares in `Undelegate` (we cap in `handleMsgBeginRedelegate ` but not here)

Reply:
@rigelrozanski yes I can update this, but is it an issue if say as a delegator I provided way more tokens than I actually have? The handler will cap at my total shares which is fine sure, but it's somewhat not obvious. I think it'd be wiser to return an error instead (which is currently does).